taskchampion/taskchampion/src/server/cloud/server.rs

```
@@ -0,0 +1,1130 @@
+    v.as_simple().to_string().into_bytes()
+}
+
+impl<SVC: Service> CloudServer<SVC> {
+    pub(in crate::server) fn new(service: SVC, encryption_secret: Vec<u8>) -> Result<Self> {
+        let cryptor = Cryptor::new(b"Taskchampion", &encryption_secret.into())?;
+        Ok(Self {
+            service,
+            cryptor,
+            cleanup_probability: DEFAULT_CLEANUP_PROBABILITY,
+            #[cfg(test)]
+            add_version_intercept: None,
+        })
+    }
+
+    /// Generate an object name for the given parent and child versions.
+    fn version_name(parent_version_id: &VersionId, child_version_id: &VersionId) -> Vec<u8> {
+        format!(
+            "v-{}-{}",
+            parent_version_id.as_simple(),
+            child_version_id.as_simple()
+        )
+        .into_bytes()
+    }
+
+    /// Parse a version name as generated by `version_name`.
+    fn parse_version_name(name: &[u8]) -> Option<(VersionId, VersionId)> {
+        if name.len() != 2 + 32 + 1 + 32 || !name.starts_with(b"v-") || name[2 + 32] != b'-' {
+            return None;
+        }
+        let Ok(parent_version_id) = VersionId::try_parse_ascii(&name[2..2 + 32]) else {
+            return None;
+        };
+        let Ok(child_version_id) = VersionId::try_parse_ascii(&name[2 + 32 + 1..]) else {
+            return None;
+        };
+        Some((parent_version_id, child_version_id))
+    }
+
+    /// Generate an object name for a snapshot at the given version.
+    fn snapshot_name(version_id: &VersionId) -> Vec<u8> {
+        format!("s-{}", version_id.as_simple()).into_bytes()
+    }
+
+    /// Parse a snapshot name as generated by `snapshot_name`.
+    fn parse_snapshot_name(name: &[u8]) -> Option<VersionId> {
+        if name.len() != 2 + 32 || !name.starts_with(b"s-") {
+            return None;
+        }
+        let Ok(version_id) = VersionId::try_parse_ascii(&name[2..2 + 32]) else {
+            return None;
+        };
+        Some(version_id)
+    }
```

Would it make sense to log a warning before each place we `return None` here? In practice it looks like only valid snapshot names should be fed into this function and I'm not sure parsing should fail silently.

---

Same